// @microsoft/applicationinsights-core-js/dist-esm/JavaScriptSDK.Enums/LoggingEnums.js
/*
 * Application Insights JavaScript SDK - Core, 2.8.1
 * Copyright (c) Microsoft and contributors. All rights reserved.
 */


import { createEnumStyle } from "../JavaScriptSDK.Enums/EnumHelperFuncs";
export var LoggingSeverity = createEnumStyle({
    CRITICAL: 1 /* CRITICAL */,
    WARNING: 2 /* WARNING */
});
/**
 * Internal message ID. Please create a new one for every conceptually different message. Please keep alphabetically ordered
 */
export var _InternalMessageId = createEnumStyle({
    // Non user actionable
    BrowserDoesNotSupportLocalStorage: 0 /* BrowserDoesNotSupportLocalStorage */,
    BrowserCannotReadLocalStorage: 1 /* BrowserCannotReadLocalStorage */,
    BrowserCannotReadSessionStorage: 2 /* BrowserCannotReadSessionStorage */,
    BrowserCannotWriteLocalStorage: 3 /* BrowserCannotWriteLocalStorage */,
    BrowserCannotWriteSessionStorage: 4 /* BrowserCannotWriteSessionStorage */,
    BrowserFailedRemovalFromLocalStorage: 5 /* BrowserFailedRemovalFromLocalStorage */,
    BrowserFailedRemovalFromSessionStorage: 6 /* BrowserFailedRemovalFromSessionStorage */,
    CannotSendEmptyTelemetry: 7 /* CannotSendEmptyTelemetry */,
    ClientPerformanceMathError: 8 /* ClientPerformanceMathError */,
    ErrorParsingAISessionCookie: 9 /* ErrorParsingAISessionCookie */,
    ErrorPVCalc: 10 /* ErrorPVCalc */,
    ExceptionWhileLoggingError: 11 /* ExceptionWhileLoggingError */,
    FailedAddingTelemetryToBuffer: 12 /* FailedAddingTelemetryToBuffer */,
    FailedMonitorAjaxAbort: 13 /* FailedMonitorAjaxAbort */,
    FailedMonitorAjaxDur: 14 /* FailedMonitorAjaxDur */,
    FailedMonitorAjaxOpen: 15 /* FailedMonitorAjaxOpen */,
    FailedMonitorAjaxRSC: 16 /* FailedMonitorAjaxRSC */,
    FailedMonitorAjaxSend: 17 /* FailedMonitorAjaxSend */,
    FailedMonitorAjaxGetCorrelationHeader: 18 /* FailedMonitorAjaxGetCorrelationHeader */,
    FailedToAddHandlerForOnBeforeUnload: 19 /* FailedToAddHandlerForOnBeforeUnload */,
    FailedToSendQueuedTelemetry: 20 /* FailedToSendQueuedTelemetry */,
    FailedToReportDataLoss: 21 /* FailedToReportDataLoss */,
    FlushFailed: 22 /* FlushFailed */,
    MessageLimitPerPVExceeded: 23 /* MessageLimitPerPVExceeded */,
    MissingRequiredFieldSpecification: 24 /* MissingRequiredFieldSpecification */,
    NavigationTimingNotSupported: 25 /* NavigationTimingNotSupported */,
    OnError: 26 /* OnError */,
    SessionRenewalDateIsZero: 27 /* SessionRenewalDateIsZero */,
    SenderNotInitialized: 28 /* SenderNotInitialized */,
    StartTrackEventFailed: 29 /* StartTrackEventFailed */,
    StopTrackEventFailed: 30 /* StopTrackEventFailed */,
    StartTrackFailed: 31 /* StartTrackFailed */,
    StopTrackFailed: 32 /* StopTrackFailed */,
    TelemetrySampledAndNotSent: 33 /* TelemetrySampledAndNotSent */,
    TrackEventFailed: 34 /* TrackEventFailed */,
    TrackExceptionFailed: 35 /* TrackExceptionFailed */,
    TrackMetricFailed: 36 /* TrackMetricFailed */,
    TrackPVFailed: 37 /* TrackPVFailed */,
    TrackPVFailedCalc: 38 /* TrackPVFailedCalc */,
    TrackTraceFailed: 39 /* TrackTraceFailed */,
    TransmissionFailed: 40 /* TransmissionFailed */,
    FailedToSetStorageBuffer: 41 /* FailedToSetStorageBuffer */,
    FailedToRestoreStorageBuffer: 42 /* FailedToRestoreStorageBuffer */,
    InvalidBackendResponse: 43 /* InvalidBackendResponse */,
    FailedToFixDepricatedValues: 44 /* FailedToFixDepricatedValues */,
    InvalidDurationValue: 45 /* InvalidDurationValue */,
    TelemetryEnvelopeInvalid: 46 /* TelemetryEnvelopeInvalid */,
    CreateEnvelopeError: 47 /* CreateEnvelopeError */,
    // User actionable
    CannotSerializeObject: 48 /* CannotSerializeObject */,
    CannotSerializeObjectNonSerializable: 49 /* CannotSerializeObjectNonSerializable */,
    CircularReferenceDetected: 50 /* CircularReferenceDetected */,
    ClearAuthContextFailed: 51 /* ClearAuthContextFailed */,
    ExceptionTruncated: 52 /* ExceptionTruncated */,
    IllegalCharsInName: 53 /* IllegalCharsInName */,
    ItemNotInArray: 54 /* ItemNotInArray */,
    MaxAjaxPerPVExceeded: 55 /* MaxAjaxPerPVExceeded */,
    MessageTruncated: 56 /* MessageTruncated */,
    NameTooLong: 57 /* NameTooLong */,
    SampleRateOutOfRange: 58 /* SampleRateOutOfRange */,
    SetAuthContextFailed: 59 /* SetAuthContextFailed */,
    SetAuthContextFailedAccountName: 60 /* SetAuthContextFailedAccountName */,
    StringValueTooLong: 61 /* StringValueTooLong */,
    StartCalledMoreThanOnce: 62 /* StartCalledMoreThanOnce */,
    StopCalledWithoutStart: 63 /* StopCalledWithoutStart */,
    TelemetryInitializerFailed: 64 /* TelemetryInitializerFailed */,
    TrackArgumentsNotSpecified: 65 /* TrackArgumentsNotSpecified */,
    UrlTooLong: 66 /* UrlTooLong */,
    SessionStorageBufferFull: 67 /* SessionStorageBufferFull */,
    CannotAccessCookie: 68 /* CannotAccessCookie */,
    IdTooLong: 69 /* IdTooLong */,
    InvalidEvent: 70 /* InvalidEvent */,
    FailedMonitorAjaxSetRequestHeader: 71 /* FailedMonitorAjaxSetRequestHeader */,
    SendBrowserInfoOnUserInit: 72 /* SendBrowserInfoOnUserInit */,
    PluginException: 73 /* PluginException */,
    NotificationException: 74 /* NotificationException */,
    SnippetScriptLoadFailure: 99 /* SnippetScriptLoadFailure */,
    InvalidInstrumentationKey: 100 /* InvalidInstrumentationKey */,
    CannotParseAiBlobValue: 101 /* CannotParseAiBlobValue */,
    InvalidContentBlob: 102 /* InvalidContentBlob */,
    TrackPageActionEventFailed: 103 /* TrackPageActionEventFailed */,
    FailedAddingCustomDefinedRequestContext: 104 /* FailedAddingCustomDefinedRequestContext */,
    InMemoryStorageBufferFull: 105 /* InMemoryStorageBufferFull */
});
